Fix DropdownButtonFormField icon not vertically centered

bleroux opened the pull request:
## Description

This PR fixes `DropdownButtonFormField` suffix icon vertical alignment when `DropdownButtonFormField.inputDecoration` is provided without a hint text.

### Before:

The dropdown icon is misaligned vertically:

![image](https://github.com/user-attachments/assets/b79b35ae-c333-4cb4-91f7-5c9ce8edd975)


### After:

The dropdown icon is aligned vertically:

![image](https://github.com/user-attachments/assets/b5704453-66a3-4cb2-97ff-bef556bff377)


## Related Issue

Fixes [DropdownButtonFormField arrow icon is misaligned vertically](https://github.com/flutter/flutter/issues/157074) 

## Tests

Adds 1 test.

bleroux committed:
Fix DropdownButtonFormField icon not vertically centered

justinmc reviewed (approved):
It's really good that https://github.com/flutter/flutter/pull/159975 simplified the solution here.

LGTM but let's see if the Google tests pass.

bleroux committed:
Replace _ in test

justinmc commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -1609,7 +1610,12 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
+      InputDecoration effectiveDecoration = widget._inputDecoration!.copyWith(
+        // Override the suffix icon constraints to allow the
+        // icon alignment to match the regular dropdown button.
+        suffixIconConstraints: const BoxConstraints(minWidth: 40.0),
```
It looks like there are some Google failures where some text that used to fit inside of the dropdown no longer fits, as if the horizontal space available in the field has been reduced. Is that due to this minWidth constraints change? Is that horizontal space change intentional?

bleroux replied:
Oh, interesting!

> as if the horizontal space available in the field has been reduced

I don't remember why we choose to set the minWidth to 40.0. Maybe something from a previous version of the PR.
I updated the PR to set minWidth to 24.0. But it won't be enough to get the same available horizontal space for the input compared to before this PR. 

The reason is that for InputDecoration there is 4.0 default gap between the input and the suffixIcon.
Previous version of the DropdownButtonFormField did not rely on InputDecoration.suffixIcon so it was not impacted by this gap.

I added an experimental commit to this PR to make it possible to remove this gap. This will not be the final solution, it is only a temporary commit to see if it helps with the Google testing failures.

bleroux replied:
@justinmc Are there many Google testing failures now?
The PR is in an experimental state (see my previous comment). The number and the kind of Google testing failures will determine how to proceed there. Thanks for having a look when you will get bandwidth (no hurry there).

justinmc replied:
I don't see any failures of the sort that I mentioned, where text no longer fits. In fact, I see some case where more text now fits than did before. However, the suffix is very close to the right edge now.

Sorry I missed this notification, feel free to ping me if I disappear again :)

bleroux commented:
Many thanks for the feedback 🙏 .
Based on https://github.com/flutter/flutter/pull/163205#discussion_r2049746555. I will mark this PR as draft, file a separate PR to make the inputDecoration prefix and suffix gaps configurable. And then come back to this PR. Especially trying to understand in which cases "the suffix is very close to the right edge now".

justinmc commented:
Here's what I meant by the suffix icon being close to the right side:

| The master branch | Your screenshot of this PR in the description | This PR currently |
| --- | --- | --- |
| ![Screenshot 2025-04-18 at 11 23 33 AM](https://github.com/user-attachments/assets/8e9495a2-86d1-48f6-8f3a-5160cdaf8172) | ![412859218-b5704453-66a3-4cb2-97ff-bef556bff377](https://github.com/user-attachments/assets/becc6fdd-66b4-4c68-a5fd-8aa0c800f7b8) | ![Screenshot 2025-04-18 at 11 22 13 AM](https://github.com/user-attachments/assets/23044fd8-a7c5-4a4e-b758-ce3af92d3579) |

You can see in the PR currently the suffix icon is further to the right than it was before you changed the PR.

<details>

<summary>Code</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Center(
              child: SizedBox(
                width: 200,
                child: DropdownButtonFormField<String>(
                  onChanged: (_) {},
                  decoration: InputDecoration(
                    fillColor: Colors.blueGrey.withOpacity(0.2),
                    filled: true,
                    labelText: 'Label text',
                  ),
                  items: [
                    DropdownMenuItem(
                      value: 'selected',
                      child: Text('Selected item'),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

bleroux commented:
Wow, thanks a lot for providing this sample and the screenshot.
For the moment, I'm surprised that the screenshot in my PR description did not match the result you got. I will try to find why there is such a discrepancy (maybe related to another InputDecoration parameter or maybe related to the platform used to run the sample). Anyway, thanks a lot for providing these very useful feedback 🙏

bleroux committed:
Change suffix icon constraints

bleroux commented:
@justinmc I updated the PR based on your last [comment](https://github.com/flutter/flutter/pull/163205#issuecomment-2815983185) and our discussion on https://github.com/flutter/flutter/pull/167558#issuecomment-2829544119

It turned out that the fix, in order to compute the correct suffixIconConstraints, should take into account if the input decoration is filled or outlined.

Because for a filled or outlined decoration the right padding is directly provided by the constraints added to the suffix icon, for instance constraints on mobile are 48x48 which means a left and right padding of 12.0. (FYI, on desktop it is be 8.0 because the suffixIconConstraints is 40x40, this is probably not intended as from the M3 spec, visual density usually applies only on the vertical direction, I will check that and file an issue if needed). 

I updated the fix accordingly. Let me know how it goes with the Google tests.

And let me try to explain the layout:

### Before the fix

DropdownButtonFormField was positioning the icon by setting the input decoration content to a Row containing both the selected item label and the icon (24x24).

For a filled or outlined decoration, the horizontal layout was: leftPadding (16.0) | selected item label | icon (24.0) | rightPadding (16.0). 
(16.0 padding since very recently because it is due to https://github.com/flutter/flutter/pull/162157, before it was 12.0).

For a non-filled and non-outlined decoration, it was: leftPadding (0.0) | selected item label | icon (24.0) | rightPadding (0.0)

### With the fix

DropdownButtonFormField is positioning the icon by setting the input decoration suffixIcon and suffixIconConstraints.

For a filled or outlined decoration, the horizontal layout is: leftPadding (16.0) | selected item label | suffixIconGap (4.0) | icon (24.0) | rightPadding (12.0). 
Here the right padding is 12.0 and this is the compliant padding from the M3 spec: when there is no suffix icon right padding should be 16.0, when the is a suffixIcon it should be 12.0). 
This means that, with the fix, the content's width is the same as before this fix. The difference is the icon moving by 4.0 pixel to the right (which will restore its position as before https://github.com/flutter/flutter/pull/162157).

For a non-filled and non-outlined decoration, it is: leftPadding (0.0) | selected item label | icon (24.0) | suffixIconGap (4.0) | rightPadding (0.0)
Here the difference will be the content width being reduced by 4.0 pixel and the icon should be at the same horizontal position.


Here is combined screenshot to illustrate this. The top 3 fields showcase the fix, the bottom ones were captured without the fix:

![image](https://github.com/user-attachments/assets/0f85dfbc-381a-43ae-abfb-a9ebf570b280)

<details><summary>Code sample for the screenshot</summary>

```dart

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  final bool filled = true;
  static const List<DropdownMenuItem<String>> items = [
    DropdownMenuItem(value: 'selected', child: Text('Selected item')),
  ];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: SizedBox(
            width: 200,
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              spacing: 4.0,
              children: [
                DropdownButtonFormField<String>(
                  onChanged: (_) {},
                  decoration: const InputDecoration(labelText: 'Raw'),
                  // Before https://github.com/flutter/flutter/pull/163205, width could be 176.
                  selectedItemBuilder:
                      (context) => [Container(height: 24, width: 172, color: Colors.amber)],
                  items: items,
                ),
                DropdownButtonFormField<String>(
                  onChanged: (_) {},
                  decoration: const InputDecoration(labelText: 'Filled', filled: true),
                  selectedItemBuilder:
                      (context) => [Container(height: 24, width: 144, color: Colors.amber)],
                  items: items,
                ),
                DropdownButtonFormField<String>(
                  onChanged: (_) {},
                  decoration: const InputDecoration(
                    labelText: 'Outlined',
                    border: OutlineInputBorder(),
                  ),
                  selectedItemBuilder:
                      (context) => [Container(height: 24, width: 144, color: Colors.amber)],
                  items: items,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

``` 
</details>

justinmc commented:
Nice, I ran it locally and confirmed that the layout seems to be correct as you explained. The Google test check has disappeared though, let me try re-adding it.

justinmc reviewed (commented):
I'm still seeing a horizontal change in the position of the suffix icon with this PR, I think that's not intentional because I don't see a horizontal change in the comparison image that you posted above in https://github.com/flutter/flutter/pull/163205#issuecomment-2830584566, right? I think this is happening only in outlined inputs. Repro code below.

Otherwise all of the other Google test failures look expected (the arrow moved slightly vertically).

| Before this PR | After this PR |
| --- | --- |
| <img src="https://github.com/user-attachments/assets/ff2a0072-db17-4793-a956-1c58987a30f7" /> |<img src="https://github.com/user-attachments/assets/e0c4057f-dba2-41c9-a5eb-7b108a38b601" /> |

<details>

<summary>Code</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Center(
              child: SizedBox(
                width: 200,
                child: DropdownButtonFormField<String>(
                  onChanged: (_) {},
                  decoration: InputDecoration(
                    border: OutlineInputBorder(),
                    labelText: 'Label text',
                  ),
                  items: [
                    DropdownMenuItem(
                      value: 'selected',
                      child: Text('Selected item'),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

```

</details>

justinmc reviewed (approved):
Renewing my LGTM :+1: . I reviewed all Google test failures and they are expected. There was some trouble with Frob but it's working now, this can be autosubmitted.
